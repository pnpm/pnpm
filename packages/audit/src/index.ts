import PnpmError from '@pnpm/error'
import fetch, { RetryOpts } from '@pnpm/fetch'
import { Lockfile } from '@pnpm/lockfile-types'
import { DependenciesField } from '@pnpm/types'
import lockfileToAuditTree from './lockfileToAuditTree'
import { AuditReport } from './types'

export * from './types'

export default async function audit (
  lockfile: Lockfile,
  opts: {
    include?: { [dependenciesField in DependenciesField]: boolean },
    registry: string,
    retry?: RetryOpts,
  },
) {
  const auditTree = lockfileToAuditTree(lockfile, { include: opts.include })
  const registry = opts.registry.endsWith('/') ? opts.registry : `${opts.registry}/`
  const auditUrl = `${registry}-/npm/v1/security/audits`
  const res = await fetch(auditUrl, {
    body: JSON.stringify(auditTree),
    headers: { 'Content-Type': 'application/json' },
    method: 'post',
    retry: opts.retry,
  })
  if (res.status !== 200) {
    throw new PnpmError('AUDIT_BAD_RESPONSE', `The audit endpoint (at ${auditUrl}) responded with ${res.status}: ${await res.text()}`)
  }
  return res.json() as Promise<AuditReport>
}
