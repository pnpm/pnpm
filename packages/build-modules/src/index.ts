import { ENGINE_NAME } from '@pnpm/constants'
import { skippedOptionalDependencyLogger } from '@pnpm/core-loggers'
import { runPostinstallHooks } from '@pnpm/lifecycle'
import linkBins, { linkBinsOfPackages } from '@pnpm/link-bins'
import logger from '@pnpm/logger'
import { fromDir as readPackageFromDir } from '@pnpm/read-package-json'
import { StoreController } from '@pnpm/store-controller-types'
import { DependencyManifest, PackageManifest } from '@pnpm/types'
import graphSequencer = require('graph-sequencer')
import path = require('path')
import R = require('ramda')
import runGroups from 'run-groups'

export default async (
  depGraph: DependenciesGraph,
  rootDepPaths: string[],
  opts: {
    childConcurrency?: number,
    depsToBuild?: Set<string>,
    extraBinPaths?: string[],
    optional: boolean,
    prefix: string,
    rawNpmConfig: object,
    unsafePerm: boolean,
    userAgent: string,
    sideEffectsCacheWrite: boolean,
    storeController: StoreController,
    rootNodeModulesDir: string,
  },
) => {
  const warn = (message: string) => logger.warn({ message, prefix: opts.prefix })
  // postinstall hooks
  const nodesToBuild = new Set<string>()
  getSubgraphToBuild(depGraph, rootDepPaths, nodesToBuild, new Set<string>())
  const onlyFromBuildGraph = R.filter((depPath: string) => nodesToBuild.has(depPath))

  const nodesToBuildArray = Array.from(nodesToBuild)
  const graph = new Map(
    nodesToBuildArray
      .map((depPath) => [depPath, onlyFromBuildGraph(R.values(depGraph[depPath].children))]) as Array<[string, string[]]>,
  )
  const graphSequencerResult = graphSequencer({
    graph,
    groups: [nodesToBuildArray],
  })
  const chunks = graphSequencerResult.chunks as string[][]
  const buildDepOpts = { ...opts, warn }
  const groups = chunks.map((chunk) => {
    chunk = chunk.filter((depPath) => depGraph[depPath].requiresBuild && !depGraph[depPath].isBuilt)
    if (opts.depsToBuild) {
      chunk = chunk.filter((depPath) => opts.depsToBuild!.has(depPath))
    }

    return chunk.map((depPath: string) =>
      async () => buildDependency(depPath, depGraph, buildDepOpts)
    )
  })
  await runGroups(opts.childConcurrency || 4, groups)
}

async function buildDependency (
  depPath: string,
  depGraph: DependenciesGraph,
  opts: {
    extraBinPaths?: string[],
    optional: boolean,
    prefix: string,
    rawNpmConfig: object,
    rootNodeModulesDir: string,
    sideEffectsCacheWrite: boolean,
    storeController: StoreController,
    unsafePerm: boolean,
    warn: (message: string) => void,
  }
) {
  const depNode = depGraph[depPath]
  try {
    await linkBinsOfDependencies(depNode, depGraph, opts)
    const hasSideEffects = await runPostinstallHooks({
      depPath,
      extraBinPaths: opts.extraBinPaths,
      optional: depNode.optional,
      pkgRoot: depNode.peripheralLocation,
      prepare: depNode.prepare,
      rawNpmConfig: opts.rawNpmConfig,
      rootNodeModulesDir: opts.rootNodeModulesDir,
      unsafePerm: opts.unsafePerm || false,
    })
    if (hasSideEffects && opts.sideEffectsCacheWrite) {
      try {
        await opts.storeController.upload(depNode.peripheralLocation, {
          engine: ENGINE_NAME,
          packageId: depNode.packageId,
        })
      } catch (err) {
        if (err && err.statusCode === 403) {
          logger.warn({
            message: `The store server disabled upload requests, could not upload ${depNode.packageId}`,
            prefix: opts.prefix,
          })
        } else {
          logger.warn({
            error: err,
            message: `An error occurred while uploading ${depNode.packageId}`,
            prefix: opts.prefix,
          })
        }
      }
    }
  } catch (err) {
    if (depNode.optional) {
      // TODO: add parents field to the log
      const pkg = await readPackageFromDir(path.join(depNode.peripheralLocation)) as DependencyManifest
      skippedOptionalDependencyLogger.debug({
        details: err.toString(),
        package: {
          id: depNode.packageId,
          name: pkg.name,
          version: pkg.version,
        },
        prefix: opts.prefix,
        reason: 'build_failure',
      })
      return
    }
    throw err
  }
}

function getSubgraphToBuild (
  graph: DependenciesGraph,
  entryNodes: string[],
  nodesToBuild: Set<string>,
  walked: Set<string>,
) {
  let currentShouldBeBuilt = false
  for (const depPath of entryNodes) {
    if (!graph[depPath]) return // packages that are already in node_modules are skipped
    if (nodesToBuild.has(depPath)) {
      currentShouldBeBuilt = true
    }
    if (walked.has(depPath)) continue
    walked.add(depPath)
    const childShouldBeBuilt = getSubgraphToBuild(graph, R.values(graph[depPath].children), nodesToBuild, walked)
      || graph[depPath].requiresBuild
    if (childShouldBeBuilt) {
      nodesToBuild.add(depPath)
      currentShouldBeBuilt = true
    }
  }
  return currentShouldBeBuilt
}

export interface DependenciesGraphNode {
  fetchingBundledManifest?: () => Promise<PackageManifest>,
  hasBundledDependencies: boolean,
  peripheralLocation: string,
  children: {[alias: string]: string},
  optional: boolean,
  optionalDependencies: Set<string>,
  packageId: string, // TODO: this option is currently only needed when running postinstall scripts but even there it should be not used
  installable?: boolean,
  isBuilt?: boolean,
  requiresBuild?: boolean,
  prepare: boolean,
  hasBin: boolean,
}

export interface DependenciesGraph {
  [depPath: string]: DependenciesGraphNode
}

export async function linkBinsOfDependencies (
  depNode: DependenciesGraphNode,
  depGraph: DependenciesGraph,
  opts: {
    optional: boolean,
    warn: (message: string) => void,
  },
) {
  const childrenToLink = opts.optional
    ? depNode.children
    : Object.keys(depNode.children)
      .reduce((nonOptionalChildren, childAlias) => {
        if (!depNode.optionalDependencies.has(childAlias)) {
          nonOptionalChildren[childAlias] = depNode.children[childAlias]
        }
        return nonOptionalChildren
      }, {})

  const binPath = path.join(depNode.peripheralLocation, 'node_modules', '.bin')

  const pkgs = await Promise.all(
    Object.keys(childrenToLink)
      .filter((alias) => {
        const dep = depGraph[childrenToLink[alias]]
        if (!dep) {
          // TODO: Try to reproduce this issue with a test in supi
          logger.debug({ message: `Failed to link bins of "${alias}" to "${binPath}". This is probably not an issue.` })
          return false
        }
        return dep.hasBin && dep.installable !== false
      })
      .map(async (alias) => {
        const dep = depGraph[childrenToLink[alias]]
        return {
          location: dep.peripheralLocation,
          manifest: dep.fetchingBundledManifest && (await dep.fetchingBundledManifest()) || (await readPackageFromDir(dep.peripheralLocation) as DependencyManifest),
        }
      }),
  )

  await linkBinsOfPackages(pkgs, binPath, { warn: opts.warn })

  // link also the bundled dependencies` bins
  if (depNode.hasBundledDependencies) {
    const bundledModules = path.join(depNode.peripheralLocation, 'node_modules')
    await linkBins(bundledModules, binPath, { warn: opts.warn })
  }
}
