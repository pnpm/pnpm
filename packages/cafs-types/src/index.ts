import type { IntegrityLike } from 'ssri'
import type { DependencyManifest } from '@pnpm/types'

export interface DeferredManifestPromise {
  resolve: (manifest: DependencyManifest | undefined) => void
  reject: (err: Error) => void
}

export interface PackageFileInfo {
  checkedAt?: number // Nullable for backward compatibility
  integrity: string
  mode: number
  size: number
}

export type PackageFilesResponse = {
  fromStore: boolean
  packageImportMethod?: 'auto' | 'hardlink' | 'copy' | 'clone' | 'clone-or-copy'
  sideEffects?: Record<string, Record<string, PackageFileInfo>>
} & ({
  local: true
  filesIndex: Record<string, string>
} | {
  local?: false
  filesIndex: Record<string, PackageFileInfo>
})

export interface ImportPackageOpts {
  requiresBuild?: boolean
  sideEffectsCacheKey?: string
  filesResponse: PackageFilesResponse
  force: boolean
}

export type ImportPackageFunction = (
  to: string,
  opts: ImportPackageOpts
) => Promise<{ isBuilt: boolean, importMethod: undefined | string }>

export type FileType = 'exec' | 'nonexec' | 'index'

export interface FilesIndex {
  [filename: string]: {
    mode: number
    size: number
    writeResult: Promise<FileWriteResult>
  }
}

export interface FileWriteResult {
  checkedAt: number
  integrity: IntegrityLike
}

export interface Cafs {
  cafsDir: string
  addFilesFromDir: (dir: string, manifest?: DeferredManifestPromise) => Promise<FilesIndex>
  addFilesFromTarball: (stream: NodeJS.ReadableStream, manifest?: DeferredManifestPromise) => Promise<FilesIndex>
  getFilePathInCafs: (integrity: string | IntegrityLike, fileType: FileType) => string
  getFilePathByModeInCafs: (integrity: string | IntegrityLike, mode: number) => string
  importPackage: ImportPackageFunction
  tempDir: () => Promise<string>
}
