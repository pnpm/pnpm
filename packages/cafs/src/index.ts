import { promises as fs, Stats } from 'fs'
import path from 'path'
import { FileWriteResult, PackageFileInfo } from '@pnpm/fetcher-base'
import getStream from 'get-stream'
import pathTemp from 'path-temp'
import renameOverwrite from 'rename-overwrite'
import ssri from 'ssri'
import addFilesFromDir from './addFilesFromDir'
import addFilesFromTarball from './addFilesFromTarball'
import checkFilesIntegrity, {
  PackageFilesIndex,
  verifyFileIntegrity,
} from './checkFilesIntegrity'
import getFilePathInCafs, {
  contentPathFromHex,
  FileType,
  getFilePathByModeInCafs,
  modeIsExecutable,
} from './getFilePathInCafs'
import writeFile from './writeFile'

export {
  checkFilesIntegrity,
  FileType,
  getFilePathByModeInCafs,
  getFilePathInCafs,
  PackageFileInfo,
  PackageFilesIndex,
}

export default function createCafs (cafsDir: string, ignore?: ((filename: string) => Boolean)) {
  const locker = new Map()
  const _writeBufferToCafs = writeBufferToCafs.bind(null, locker, cafsDir)
  const addStream = addStreamToCafs.bind(null, _writeBufferToCafs)
  const addBuffer = addBufferToCafs.bind(null, _writeBufferToCafs)
  return {
    addFilesFromDir: addFilesFromDir.bind(null, { addBuffer, addStream }),
    addFilesFromTarball: addFilesFromTarball.bind(null, addStream, ignore ?? null),
  }
}

async function addStreamToCafs (
  writeBufferToCafs: WriteBufferToCafs,
  fileStream: NodeJS.ReadableStream,
  mode: number
): Promise<FileWriteResult> {
  const buffer = await getStream.buffer(fileStream)
  return addBufferToCafs(writeBufferToCafs, buffer, mode)
}

type WriteBufferToCafs = (buffer: Buffer, fileDest: string, mode: number | undefined, integrity: ssri.IntegrityLike) => Promise<number>

async function addBufferToCafs (
  writeBufferToCafs: WriteBufferToCafs,
  buffer: Buffer,
  mode: number
): Promise<FileWriteResult> {
  const integrity = ssri.fromData(buffer)
  const isExecutable = modeIsExecutable(mode)
  const fileDest = contentPathFromHex(isExecutable ? 'exec' : 'nonexec', integrity.hexDigest())
  const checkedAt = await writeBufferToCafs(
    buffer,
    fileDest,
    isExecutable ? 0o755 : undefined,
    integrity
  )
  return { checkedAt, integrity }
}

async function writeBufferToCafs (
  locker: Map<string, Promise<number>>,
  cafsDir: string,
  buffer: Buffer,
  fileDest: string,
  mode: number | undefined,
  integrity: ssri.IntegrityLike
): Promise<number> {
  fileDest = path.join(cafsDir, fileDest)
  if (locker.has(fileDest)) {
    return locker.get(fileDest)!
  }
  const p = (async () => {
    // This part is a bit redundant.
    // When a file is already used by another package,
    // we probably have validated its content already.
    // However, there is no way to find which package index file references
    // the given file. So we should revalidate the content of the file again.
    if (await existsSame(fileDest, integrity)) {
      return Date.now()
    }

    // This might be too cautious.
    // The write is atomic, so in case pnpm crashes, no broken file
    // will be added to the store.
    // It might be a redundant step though, as we verify the contents of the
    // files before linking
    //
    // If we don't allow --no-verify-store-integrity then we probably can write
    // to the final file directly.
    const temp = pathTemp(path.dirname(fileDest))
    await writeFile(temp, buffer, mode)
    // Unfortunately, "birth time" (time of file creation) is available not on all filesystems.
    // We log the creation time ourselves and save it in the package index file.
    // Having this information allows us to skip content checks for files that were not modified since "birth time".
    const birthtimeMs = Date.now()
    await renameOverwrite(temp, fileDest)
    return birthtimeMs
  })()
  locker.set(fileDest, p)
  return p
}

async function existsSame (filename: string, integrity: ssri.IntegrityLike) {
  let existingFile: Stats | undefined
  try {
    existingFile = await fs.stat(filename)
  } catch (err) {
    return false
  }
  return verifyFileIntegrity(filename, {
    size: existingFile.size,
    integrity,
  })
}
