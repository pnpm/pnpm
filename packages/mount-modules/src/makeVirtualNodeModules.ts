import path from 'path'
import { Lockfile } from '@pnpm/lockfile-file'
import { DEPENDENCIES_FIELDS } from '@pnpm/types'
import { nameVerFromPkgSnapshot } from '@pnpm/lockfile-utils'
import * as dp from 'dependency-path'
import normalize from 'normalize-path'

type DirEntry = {
  entryType: 'index'
  depPath: string
} | {
  entryType: 'symlink'
  target: string
} | {
  entryType: 'directory'
  entries: Record<string, DirEntry>
}

export default function (lockfile: Lockfile, lockfileDir: string): DirEntry {
  const entries: Record<string, DirEntry> = {
    '.pnpm': {
      entryType: 'directory',
      entries: createVirtualStoreDir(lockfile, lockfileDir),
    },
  }
  for (const depType of DEPENDENCIES_FIELDS) {
    for (const [depName, ref] of Object.entries(lockfile.importers['.'][depType] ?? {})) {
      const symlink: DirEntry = {
        entryType: 'symlink',
        target: `./.pnpm/${dp.depPathToFilename(dp.refToRelative(ref, depName)!, lockfileDir)}/node_modules/${depName}`,
      }
      addDirEntry(entries, depName, symlink)
    }
  }
  return {
    entryType: 'directory',
    entries,
  }
}

function createVirtualStoreDir (lockfile: Lockfile, lockfileDir: string) {
  const rootDir = {} as Record<string, DirEntry>
  for (const [depPath, pkgSnapshot] of Object.entries(lockfile.packages ?? {})) {
    const { name } = nameVerFromPkgSnapshot(depPath, pkgSnapshot)
    const pkgNodeModules = {} as Record<string, DirEntry>
    const currentPath = dp.depPathToFilename(depPath, lockfileDir)
    const pkgDir: DirEntry = {
      entryType: 'index',
      depPath,
    }
    addDirEntry(pkgNodeModules, name, pkgDir)
    for (const [depName, ref] of Object.entries({ ...pkgSnapshot.dependencies, ...pkgSnapshot.optionalDependencies })) {
      const symlink: DirEntry = {
        entryType: 'symlink',
        target: normalize(path.relative(`${currentPath}/node_modules/`, `${dp.depPathToFilename(dp.refToRelative(ref, depName)!, lockfileDir)}/node_modules/${depName}`)),
      }
      addDirEntry(pkgNodeModules, depName, symlink)
    }
    addDirEntry(rootDir, currentPath, {
      entryType: 'directory',
      entries: {
        node_modules: {
          entryType: 'directory',
          entries: pkgNodeModules,
        },
      },
    })
  }
  return rootDir
}

function addDirEntry (target: Record<string, DirEntry>, subPath: string[] | string, newEntry: DirEntry) {
  const subPathArray = typeof subPath === 'string' ? subPath.split('/') : subPath
  const p = subPathArray.shift()!
  if (subPathArray.length) {
    if (!target[p]) {
      target[p] = {
        entryType: 'directory',
        entries: {},
      } as DirEntry
    } else if (target[p].entryType !== 'directory') {
      throw new Error()
    }
    addDirEntry(target[p]['entries'], subPathArray, newEntry)
  } else {
    target[p] = newEntry
  }
}
