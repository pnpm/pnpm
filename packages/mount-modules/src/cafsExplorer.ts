import { type PackageFilesIndex } from '@pnpm/cafs'

export function readdir (index: PackageFilesIndex, dir: string) {
  const dirs = new Set<string>()
  const prefix = dir ? `${dir}/` : ''
  for (const filePath of Object.keys(index.files)) {
    if (filePath.startsWith(prefix)) {
      const parts = filePath.substring(dir.length).split('/')
      dirs.add(parts[0] || parts[1])
    }
  }
  return Array.from(dirs)
}

export function dirEntityType (index: PackageFilesIndex, p: string) {
  if (index.files[p]) return 'file'
  const prefix = `${p}/`
  return Object.keys(index.files).some((k) => k.startsWith(prefix)) ? 'directory' : undefined
}
