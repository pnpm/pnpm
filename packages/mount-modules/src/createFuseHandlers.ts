import fs from 'fs'
import { getFilePathInCafs, getFilePathByModeInCafs, PackageFilesIndex } from '@pnpm/cafs'
import { Lockfile, readWantedLockfile, PackageSnapshot } from '@pnpm/lockfile-file'
import {
  nameVerFromPkgSnapshot,
} from '@pnpm/lockfile-utils'
import * as schemas from 'hyperdrive-schemas'
import loadJsonFile from 'load-json-file'
import Fuse from 'fuse-native'
import * as cafsExplorer from './cafsExplorer'
import makeVirtualNodeModules from './makeVirtualNodeModules'

const TIME = new Date()
const STAT_DEFAULT = {
  mtime: TIME,
  atime: TIME,
  ctime: TIME,
  nlink: 1,
  uid: process.getuid ? process.getuid() : 0,
  gid: process.getgid ? process.getgid() : 0,
}

export default async function createFuseHandlers (lockfileDir: string, cafsDir: string) {
  const lockfile = await readWantedLockfile(lockfileDir, { ignoreIncompatible: true })
  if (!lockfile) throw new Error('Cannot generate a .pnp.js without a lockfile')
  return createFuseHandlersFromLockfile(lockfile, lockfileDir, cafsDir)
}

/* eslint-disable standard/no-callback-literal */
export function createFuseHandlersFromLockfile (lockfile: Lockfile, lockfileDir: string, cafsDir: string) {
  const pkgSnapshotCache = new Map<string, { name: string, version: string, pkgSnapshot: PackageSnapshot, index: PackageFilesIndex }>()
  const virtualNodeModules = makeVirtualNodeModules(lockfile, lockfileDir)
  return {
    open (p: string, flags: string | number, cb: (exitCode: number, fd?: number) => void) {
      const dirEnt = getDirEnt(p)
      if (dirEnt?.entryType !== 'index') {
        cb(-1)
        return
      }
      const fileInfo = dirEnt.index.files[dirEnt.subPath]
      if (!fileInfo) {
        cb(-1)
        return
      }
      const filePathInStore = getFilePathByModeInCafs(cafsDir, fileInfo.integrity, fileInfo.mode)
      fs.open(filePathInStore, flags, (err, fd) => {
        if (err) {
          cb(-1)
          return
        }
        cb(0, fd)
      })
    },
    release (p: string, fd: number, cb: (exitCode: number) => void) {
      fs.close(fd, (err) => cb(err ? -1 : 0))
    },
    read (p: string, fd: number, buffer: Buffer, length: number, position: number, cb: (readBytes: number) => void) {
      fs.read(fd, buffer, position, length, position, (err, bytesRead) => {
        if (err) {
          cb(-1)
          return
        }
        cb(bytesRead)
      })
    },
    readlink (p: string, cb: (returnCode: number, target?: string) => void) {
      const dirEnt = getDirEnt(p)
      if (dirEnt?.entryType !== 'symlink') {
        cb(Fuse.ENOENT)
        return
      }
      cb(0, dirEnt.target)
    },
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    getattr (p: string, cb: (returnCode: number, files?: any) => void) {
      const dirEnt = getDirEnt(p)
      if (!dirEnt) {
        cb(Fuse.ENOENT)
        return
      }
      if (dirEnt.entryType === 'directory' || dirEnt.entryType === 'index' && !dirEnt.subPath) {
        cb(0, schemas.Stat.directory({
          ...STAT_DEFAULT,
          size: 1,
        }))
        return
      }
      if (dirEnt.entryType === 'symlink') {
        cb(0, schemas.Stat.symlink({
          ...STAT_DEFAULT,
          size: 1,
        }))
        return
      }
      if (dirEnt.entryType === 'index') {
        switch (cafsExplorer.dirEntityType(dirEnt.index, dirEnt.subPath)) {
        case 'file':
          cb(0, schemas.Stat.file({
            ...STAT_DEFAULT,
            size: dirEnt.index.files[dirEnt.subPath].size,
          }))
          return
        case 'directory':
          cb(0, schemas.Stat.directory({
            ...STAT_DEFAULT,
            size: 1,
          }))
          return
        default:
          cb(Fuse.ENOENT)
          return
        }
      }
      cb(Fuse.ENOENT)
    },
    readdir,
  }
  function readdir (p: string, cb: (returnCode: number, files?: string[]) => void) {
    const dirEnt = getDirEnt(p)
    if (dirEnt?.entryType === 'index') {
      const dirEnts = cafsExplorer.readdir(dirEnt.index, dirEnt.subPath)
      if (dirEnts.length === 0) {
        cb(Fuse.ENOENT)
        return
      }
      cb(0, dirEnts)
      return
    }
    if (!dirEnt || dirEnt.entryType !== 'directory') {
      cb(Fuse.ENOENT)
      return
    }
    cb(0, Object.keys(dirEnt.entries))
  }
  function getDirEnt (p: string) {
    let currentDirEntry = virtualNodeModules
    const parts = p === '/' ? [] : p.split('/')
    parts.shift()
    while (parts.length && currentDirEntry && currentDirEntry.entryType === 'directory') {
      currentDirEntry = currentDirEntry.entries[parts.shift()!]
    }
    if (currentDirEntry?.entryType === 'index') {
      const pkg = getPkgInfo(currentDirEntry.depPath, cafsDir)
      if (!pkg) {
        return null
      }
      return {
        ...currentDirEntry,
        index: pkg.index,
        subPath: parts.join('/'),
      }
    }
    return currentDirEntry
  }
  function getPkgInfo (depPath: string, cafsDir: string) {
    if (!pkgSnapshotCache.has(depPath)) {
      const pkgSnapshot = lockfile.packages?.[depPath]
      if (!pkgSnapshot) return undefined
      const indexPath = getFilePathInCafs(cafsDir, pkgSnapshot.resolution['integrity'], 'index')
      pkgSnapshotCache.set(depPath, {
        ...nameVerFromPkgSnapshot(depPath, pkgSnapshot),
        pkgSnapshot,
        index: loadJsonFile.sync<PackageFilesIndex>(indexPath), // TODO: maybe make it async?
      })
    }
    return pkgSnapshotCache.get(depPath)
  }
}
