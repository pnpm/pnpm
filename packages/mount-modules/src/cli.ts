import { getConfig } from '@pnpm/config'
import { promises as fs } from 'fs'
import path from 'path'

import { getStorePath } from '@pnpm/store-path'
import Fuse from 'fuse-native'
import { createFuseHandlers } from './createFuseHandlers'
(async () => { /* eslint-disable-line */
  const mnt = path.join(process.cwd(), 'node_modules')
  await fs.mkdir(mnt, { recursive: true })
  const { config } = await getConfig({
    cliOptions: {},
    packageManager: { name: '', version: '' },
  })
  const cafsDir = path.join(await getStorePath({
    pkgRoot: process.cwd(),
    storePath: config.storeDir,
    pnpmHomeDir: config.pnpmHomeDir,
  }), 'files')
  const fuse = new Fuse(mnt, await createFuseHandlers(process.cwd(), cafsDir), { debug: true })
  fuse.mount(function (err?: Error) {
    if (err != null) console.error(err)
  })

  process.once('SIGINT', function () {
    fuse.unmount((err?: Error) => {
      if (err != null) {
        console.log(`filesystem at ${fuse.mnt as string} not unmounted`, err)
      } else {
        console.log(`filesystem at ${fuse.mnt as string} unmounted`)
      }
    })
  })
})()
