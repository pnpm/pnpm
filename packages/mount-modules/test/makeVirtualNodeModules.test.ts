import path from 'path'
import { readWantedLockfile } from '@pnpm/lockfile-file'
import { makeVirtualNodeModules } from '../src/makeVirtualNodeModules'

test('makeVirtualNodeModules', async () => {
  const lockfile = await readWantedLockfile(path.join(__dirname, '__fixtures__/simple'), { ignoreIncompatible: true })
  expect(makeVirtualNodeModules(lockfile!)).toMatchSnapshot()
})
