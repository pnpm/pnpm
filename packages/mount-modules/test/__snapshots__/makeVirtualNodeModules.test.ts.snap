// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeVirtualNodeModules 1`] = `
Object {
  "entries": Object {
    ".pnpm": Object {
      "entries": Object {
        "@zkochan#git-config@0.1.0": Object {
          "entries": Object {
            "node_modules": Object {
              "entries": Object {
                "@zkochan": Object {
                  "entries": Object {
                    "git-config": Object {
                      "depPath": "/@zkochan/git-config/0.1.0",
                      "entryType": "index",
                    },
                  },
                  "entryType": "directory",
                },
                "ini": Object {
                  "entryType": "symlink",
                  "target": "../../ini@1.3.8/node_modules/ini",
                },
              },
              "entryType": "directory",
            },
          },
          "entryType": "directory",
        },
        "ini@1.3.8": Object {
          "entries": Object {
            "node_modules": Object {
              "entries": Object {
                "ini": Object {
                  "depPath": "/ini/1.3.8",
                  "entryType": "index",
                },
              },
              "entryType": "directory",
            },
          },
          "entryType": "directory",
        },
        "is-positive@1.0.0": Object {
          "entries": Object {
            "node_modules": Object {
              "entries": Object {
                "is-positive": Object {
                  "depPath": "/is-positive/1.0.0",
                  "entryType": "index",
                },
              },
              "entryType": "directory",
            },
          },
          "entryType": "directory",
        },
      },
      "entryType": "directory",
    },
    "@zkochan": Object {
      "entries": Object {
        "git-config": Object {
          "entryType": "symlink",
          "target": "./.pnpm/@zkochan#git-config@0.1.0/node_modules/@zkochan/git-config",
        },
      },
      "entryType": "directory",
    },
    "is-positive": Object {
      "entryType": "symlink",
      "target": "./.pnpm/is-positive@1.0.0/node_modules/is-positive",
    },
  },
  "entryType": "directory",
}
`;
